File: Python/Data_Preprocessing.py
# -*- coding: utf-8 -*-
"""
ML Lab for the udemy course "Machine Learning A-Z (Codes and Datasets)" 
Topic: Data Pre-Processing
@author: Anand
"""
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

from  sklearn.impute import SimpleImputer

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

from sklearn.preprocessing import LabelEncoder

from sklearn.model_selection import train_test_split

from sklearn.preprocessing import StandardScaler

dataset = pd.read_csv("./Data/Data.csv")
X = dataset.iloc[:,:-1] #Selecting all but the last columns of the dataset
Y = dataset.iloc[:,-1] #Selecting the last column of the dataset

#Handling missing data
imputer = SimpleImputer(missing_values=np.nan,strategy='mean')
imputer.fit(X.iloc[:,1:])
X.iloc[:,1:] = imputer.transform(X.iloc[:,1:])
#Additional steps to iloc while fit is added to both source and target slices

#Encoding the independend data=>Dummy Variable for categorical columns
ct = ColumnTransformer(transformers = [('encoder',OneHotEncoder(),[0])],remainder='passthrough')
X = np.array(ct.fit_transform(X))

#Encoding the dependent value
le = LabelEncoder()
Y = le.fit_transform(Y)


#Split the data into train & test set
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2, random_state=1)

'''
#Feature Scaling---applied after the slpit as otherwise there could be data leakage from the test set into the train set
#Standardisation technique ==== (x - mean(x)) / (std dev(x))
#Normlisation technique ==== (x - min(x)) / (max(x) - min(x))
Standardisation is used for general data irrecpective of the nature,
Normalisation is applied on data that is distributed normally
'''
sc = StandardScaler()
X_train[:,3:] = sc.fit_transform(X_train[:,3:])
X_test[:,3:] = sc.fit_transform(X_test[:,3:])

File: R/data_preprocessing.R
#' 
#' ML Lab for the udemy course "Machine Learning A-Z (Codes and Datasets)"
#' Topic: Data Preprocessing
#' @author: Anand


#Import the dataset#
dataset <- read.csv("../Data/Data.csv")

#Handling missing data#
dataset$Age = ifelse(is.na(dataset$Age), mean(dataset$Age, na.rm = TRUE), dataset$Age)
dataset$Salary = ifelse(is.na(dataset$Salary), mean(dataset$Salary, na.rm = TRUE), dataset$Salary)

#Encode categorical values#
levels <- unique(dataset$Country)
labels <- 1:length(levels)
dataset$Country = factor(dataset$Country, levels = levels, labels = labels)

levels <- unique(dataset$Purchased)
labels <- 0 : (length(levels) - 1)
dataset$Purchased = factor(dataset$Purchased, levels = levels, labels = labels)

#Splitting the data into train and test#
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = .8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

#Feature Scaling#
training_set[,2:3] = scale(training_set[,2:3])
test_set[,2:3] = scale(test_set[,2:3])
